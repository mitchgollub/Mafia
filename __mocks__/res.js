'use strict';

const res = {};

res.status = function (statusCode) {
    this.statusCode = statusCode
    return {
        json: (object) => {
            res.json = object
        }
    }
};

res.statusCode = 0;
res.json = {};

module.exports = res;