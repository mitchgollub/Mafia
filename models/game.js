export default class Game {
  constructor({ code, players } = {}) {
    this.code = code || '';
    this.players = players || { current: [], available: [] };
  }
}
