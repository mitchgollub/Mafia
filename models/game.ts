import Player from "./player";
import AvailableRole from "./availableRole";
import CurrentRole from "./currentRole";

export default class Game {
  constructor({ code, players }: Partial<Game> = {}) {
    this.code = code || '';
    this.players = players || { current: [], available: [] };
  }

  code: string;
  players: PlayerAssignments;

}

interface PlayerAssignments {
  current: CurrentRole[];
  available: AvailableRole[];
}
