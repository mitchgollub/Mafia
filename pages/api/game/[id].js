const db = require('../../../lib/db');
const escape = require('sql-template-strings')

export default async (req, res) => {
    console.log('id: ' + req.query.id)
    const resp = await db.query(escape`SELECT players FROM Games WHERE game_code = ${req.query.id}`)
    console.log(resp)
    if (resp && resp.length) {
        const players = JSON.parse(resp[0].players);
        res.status(200).json({ game: { code: req.query.id, players: players.current } });
    }
    
    res.status(500)
}