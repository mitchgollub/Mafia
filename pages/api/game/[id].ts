import GameRepository from '../../../repositories/gameRepository';
import GameView from '../../../views/gameView';
import { NextApiRequest, NextApiResponse } from 'next';
import { BadRequest, InternalServerError } from '../../../lib/error';

export default async (
  req: NextApiRequest,
  res: NextApiResponse<GameView | null>,
): Promise<NextApiResponse<GameView | null>> => {
  try {
    console.log(`id: ${req.query.id}`);

    const game = await GameRepository.getGame(req.query.id as string);

    if (!game) {
      return BadRequest(res, 'Could not find Game');
    }

    res.status(200).json(
      new GameView({
        code: game.code,
        players: game.players.current,
      }),
    );

    return res;
  } catch (error) {
    return InternalServerError(res, error);
  }
};
