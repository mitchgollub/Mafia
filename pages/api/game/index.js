const db = require('../../../lib/db');
const escape = require('sql-template-strings')

export default async (req, res) => {
    var result = '';
    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    var charactersLength = characters.length;
    for (var i = 0; i < 4; i++) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    console.log(req.body)
    const config = req.body;

    const cleanedConfig = cleanInput(config);

    const availableString = []
    for (let index = 0; index < cleanedConfig.Cop; index++) {
        availableString.push("Cop")
    }
    for (let index = 0; index < cleanedConfig.Mafia; index++) {
        availableString.push("Mafia")
    }
    for (let index = 0; index < cleanedConfig.Townsfolk; index++) {
        availableString.push("Townsfolk")
    }

    const available = []
    for (let index = 0; index < availableString.length; index++) {
        available.push({ id: index + 2, role: availableString[index] })
    }

    const players = {
        current: [{ id: 1, role: 'Narrator', name: 'YOU' }],
        available: available
    }

    const resp = await db.query(escape`INSERT INTO Games(game_code, players) 
        VALUES(${result}, ${JSON.stringify(players)})`)

    res.status(200).json({ game: { code: result, players: players.current } });
}

function cleanInput(config) {
    config.Cop = cleanNumber(config.Cop);
    config.Mafia = cleanNumber(config.Mafia);
    config.Townsfolk = cleanNumber(config.Townsfolk);
    return config;
}

function cleanNumber(input) {
    if (input) {
        const inputInt = parseInt(input);
        if (inputInt !== NaN && inputInt <= 10 && inputInt > 0) {
            return inputInt;
        }
    }

    return 0;
}
