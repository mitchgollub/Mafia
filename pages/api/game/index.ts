import GameRepository from '../../../repositories/gameRepository';
import { NextApiRequest, NextApiResponse } from 'next';
import { BadRequest, InternalServerError } from '../../../lib/error';
import GameView from '../../../views/gameView';
import Joi from '@hapi/joi';

const schema = Joi.array().items(
  Joi.object({
    role: Joi.string().required(),
    roleName: Joi.string().required(),
    startingValue: Joi.number().min(0).max(10).required(),
  }),
);

export default async (
  req: NextApiRequest,
  res: NextApiResponse<GameView | null>,
  gameRepository: GameRepository = new GameRepository(),
): Promise<NextApiResponse> => {
  try {
    console.log(`request: ${JSON.stringify(req.body)}`);

    const { error } = schema.validate(req.body);

    if (error) {
      return BadRequest(res, error.message);
    }

    const game = await gameRepository.createGame(req.body);

    if (!game) {
      return BadRequest(res, 'Could not create game');
    }

    res.status(200).json(
      new GameView({
        code: game.code,
        players: game.players.current,
      }),
    );
    return res;
  } catch (error) {
    return InternalServerError(res, error);
  }
};
