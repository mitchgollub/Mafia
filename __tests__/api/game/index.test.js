import res from '../../../__mocks__/res';
import game from '../../../pages/api/game/index';

test('Creates Game', async () => {
  const gameRepository = {
    createGame: jest.fn().mockResolvedValue({
      code: 'AAAA',
      players: {
        current: [{ id: 1, name: 'YOU', role: 'Narrator', session: '' }],
        available: [],
      },
    }),
  };

  const req = {
    body: [
      {
        role: 'Cop',
        roleName: 'Cops',
        startingValue: 2,
      },
    ],
  };

  const response = await game(req, res, gameRepository);

  expect(response.statusCode).toBe(200);
  expect(response.json).toMatchSnapshot();
});

test('Returns 400 when startingValues > 10', async () => {
  const gameRepository = {
    createGame: jest.fn().mockResolvedValue({
      code: 'AAAA',
      players: { current: [], available: [] },
    }),
  };

  const req = {
    body: [
      {
        role: 'Cop',
        roleName: 'Cops',
        startingValue: 12,
      },
    ],
  };

  const response = await game(req, res, gameRepository);

  expect(response.statusCode).toBe(400);
});

test('Returns 400 when null startingValues', async () => {
  const gameRepository = {
    createGame: jest.fn().mockResolvedValue({
      code: 'AAAA',
      players: { current: [], available: [] },
    }),
  };

  const req = {
    body: [
      {
        role: 'Cop',
        roleName: 'Cops',
        startingValue: null,
      },
    ],
  };

  const response = await game(req, res, gameRepository);

  expect(response.statusCode).toBe(400);
});

test('Returns 500 on error', async () => {
  const gameRepository = {
    createGame: () => {
      throw new Error('Failed to connect');
    },
  };

  const req = {
    body: [
      {
        role: 'Cop',
        roleName: 'Cops',
        startingValue: 2,
      },
    ],
  };

  const response = await game(req, res, gameRepository);

  expect(response.statusCode).toBe(500);
});
