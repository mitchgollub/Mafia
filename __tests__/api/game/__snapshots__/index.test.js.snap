// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates Game 1`] = `
GameView {
  "code": "AAAA",
  "players": Array [
    Object {
      "id": 1,
      "name": "YOU",
      "role": "Narrator",
      "session": "",
    },
  ],
}
`;
